src/gun_pool.erl

```diff
@@ -0,0 +1,600 @@
+
+%% @todo Probably don't mix Gun and pool options. Have Gun Opts in a pool option.
+init({Host, Port, Opts}) ->
+	Transport = maps:get(transport, Opts, default_transport(Port)),
+	Authority = gun_http:host_header(Transport, Host, Port),
+	Scope = maps:get(scope, Opts, default),
+	true = ets:insert_new(gun_pools, {{Scope, iolist_to_binary(Authority)}, self()}),
+	Tid = ets:new(gun_pooled_conns, [ordered_set, public]),
+	Size = maps:get(size, Opts, 8),
+	%% @todo Only start processes in static mode.
+	ConnOpts = conn_opts(Tid, Opts),
+	Conns = maps:from_list([begin
+		{ok, ConnPid} = gun:open(Host, Port, ConnOpts),
+		_ = monitor(process, ConnPid),
+		{ConnPid, down}
+	end || _ <- lists:seq(1, Size)]),
+	State = #state{
+		host=Host,
+		port=Port,
+		opts=Opts,
+		table=Tid,
+		conns=Conns
+	},
+	{ok, degraded, State}.
+
+default_transport(443) -> gun_tls;
+default_transport(_) -> gun_tcp.
+
+conn_opts(Tid, Opts0) ->
+	EventHandlerState = maps:with([event_handler], Opts0),
+	H2Opts = maps:get(http2_opts, Opts0, #{}),
+	Opts = Opts0#{
+		event_handler => {gun_pool_events_h, EventHandlerState#{
+			table => Tid
+		}},
+		http2_opts => H2Opts#{
+			notify_settings_changed => true
+		}
+	},
+	maps:without([scope, setup_fun, size], Opts).
+
+%% We use the degraded state as long as at least one connection is degraded.
+%% @todo Probably count degraded connections separately to avoid counting every time.
+degraded(info, Msg={gun_up, ConnPid, _}, StateData=#state{opts=Opts, conns=Conns}) ->
+	#{ConnPid := down} = Conns,
+	%% We optionally run the setup function if one is defined. The
+	%% setup function tells us whether we are fully up or not. The
+	%% setup function may be called repeatedly until the connection
+	%% is established.
+	%%
+	%% @todo It is possible that the connection never does get
+	%% fully established. We should deal with this. We probably
+	%% need to handle all messages.
+	{SetupFun, SetupState0} = setup_fun(Opts),
+	degraded_setup(ConnPid, Msg, StateData, SetupFun, SetupState0);
+%% @todo
+%degraded(info, Msg={gun_tunnel_up, ConnPid, _, _}, StateData0=#state{conns=Conns}) ->
+%	;
+degraded(info, Msg={gun_upgrade, ConnPid, _, _, _},
+		StateData=#state{opts=#{setup_fun := {SetupFun, _}}, conns=Conns}) ->
+	%% @todo Probably shouldn't crash if the state is incorrect, that's programmer error though.
+	#{ConnPid := {setup, SetupState0}} = Conns,
+	%% We run the setup function again using the state previously kept.
+	%% @todo We need to keep track of the StreamRef for Websocket I guess?
+	degraded_setup(ConnPid, Msg, StateData, SetupFun, SetupState0);
+degraded(Type, Event, StateData) ->
+	handle_common(Type, Event, ?FUNCTION_NAME, StateData).
+
+setup_fun(#{setup_fun := SetupFun}) ->
+	SetupFun;
+setup_fun(_) ->
+	{fun (_, {gun_up, _, Protocol}, _) ->
+		{up, Protocol, #{}}
+	end, undefined}.
+
+degraded_setup(ConnPid, Msg, StateData0=#state{conns=Conns, conns_meta=ConnsMeta}, SetupFun, SetupState0) ->
```

Comment:
An optional setup fun can be used to reestablish the Websocket connection (or later, tunnel).